### PR TITLE
Add dictionary extension

### PR DESCRIPTION
**Pull request by DrChai**
## Description

Search any words with multiple online dictionaries or translation engines in one place -- lightweight, zero dependencies.
-  Search across multiple online dictionaries and translation engines
-  Provides command `Search from Cursor Selection` to quickly look up your currently selected words. Feel free to bind a hotkey like double <kbd>⌘ cmd</kbd> for quick access.
- View related definitions and example sentences when available.
- Supports auto reverse(fallback) search, whenever the detected input matches the default language.
- Access TTS(Text-to-Speech) for the following engines:
  - Google Translation
  - Youdao
- [**Bonus**] You can configure this extension directly from the command line in an old-fashioned way!

## Screencast
dictionary-1
dictionary-4
dictionary-6

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Review by thomaslombart — COMMENTED** [review]
Hey there, could you please add a description and a screencast? That will make reviewing the PR a bit easier 😊

Thanks!

**Comment by DrChai** [comment]
@thomaslombart I just added a description and some screenshots. Apologies for not checking the actual PR comments and assuming everything was done after running `ray publish`.😥

**Review by thomaslombart — CHANGES_REQUESTED** [review]
Thanks for this PR 💪 
I've reviewed it and have some feedback.

### Configuration Screen

There's a flickering on the initial load because I haven't configured anything yet. Also, if I configure something I need to press the ESC key to go back which doesn't feel intuitive. Is it possible to get rid of that? I can see this is mainly used for the default and fallback languages. Why not moving them to the preferences? That way the screen can be removed.

https://user-images.githubusercontent.com/16003285/231463483-5e6adbcd-b20e-4196-8079-e8456bc47d91.mp4

### Empty screen

The empty screen doesn't do much when launching the command. Note that you can use the `EmptyView` component to customize the title and the description to invite the user to do something.

CleanShot 2023-04-12 at 14 48 48@2x

### Switching to metadata view

The list switches to a metadata view after the first item which feels a bit weird.

Since the first item is used to show the actions already available in the action panel, why not removing it? That way you can keep the metadata view only.

https://user-images.githubusercontent.com/16003285/231464790-df8e625d-8341-42ee-9bb1-1b9ed323aa5f.mp4

### Cleaning code

There are a bunch of commented lines and `console.debug` statements in the code. Could you remove them to make the code cleaner for future potential contributors?

---

Please, request a new review when you are ready. Feel free to contact me here or in the [Slack community](https://www.raycast.com/community) if you have any questions.

**Comment by DrChai** [comment]
@thomaslombart Thank you for taking the time and effort to review my PR. I will go through each feedback and submit a new commit this week.

File: extensions/dictionary/package.json [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  "commands": [
+    {
+      "name": "fromCmd",
+      "title": "Search in Dictionaries",
+      "description": "Search any words with multiple online dictionaries",
+      "mode": "view",
+      "arguments": [
+        {
+          "name": "arguments-title",
+          "placeholder": "-set",
```
[the remark this answers is not in the document]
Comment (DrChai) [reply extensions/dictionary/package.json:23]: tried to play with the argument and see how it works.😅 will remove it in the next commit

File: extensions/dictionary/src/engines/urban.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  const { definition, example, permalink, word, defid } = def;
+  const regex = /\[([^\]]+)\]/g;
+  const linkUrl = (term: string) => `${baseUrl}?term=${encodeURIComponent(term)}`;
+  const encode = (str: string) => str.replace(regex, (_, term) => `[${term}](${linkUrl(term)})`);
+  const markdown = `#### Definition
+## ${encode(definition)}
+#### Example
+> ${encode(example)}
+
+#### 👍   ${def.thumbs_up}   |   👎   ${def.thumbs_down}
```
[the remark this answers is not in the document]
Comment (DrChai) [reply extensions/dictionary/src/engines/urban.ts:62]: Currently, when the input query is a bit longer, accessory values would also get truncated.
Screenshot 2023-04-12 at 5 12 08 PM

**Comment by DrChai** [comment]
So I fixed the configuration view that appeared on the initial load. The issue was actually caused by an unset engine variable on the dropdown menu, it now uses Google as the engine and 'en' as the language.
 I have also considered using a metadata view, but it appears that many definitions not come with related examples:
 
Screenshot 2023-04-13 at 2 17 31 AM
There would be too much white space on the metadata view.😔

I switched to using `useCachedState`. I tried to create a custom hook that also subscribes to the cache on view mount, triggering a rerender when specific keys update.   Turns out the `useCachedState` already includes this function 👍, although it is not mentioned in the documentation. 
lastly, I added some custom messages on the emptyview:
Screenshot 2023-04-13 at 3 35 42 AM

**Comment by thomaslombart** [comment]
Thanks for addressing the issues 💪  
I've reviewed it again and have spotted other things.

### Wrong title for empty screen

The title says `The word(s) will be translated to 'en'` when using `Urban Dictionary`. As far as it looks, there is no translation with the urban dictionary, right?

CleanShot 2023-04-19 at 10 44 43@2x

Also for Google Translation, ideally, the world should correspond to the plain language. For example: `The word(s) will be translated to 'en'` would become `The word(s) will be translated to English`.

### Empty screen while searching

The empty screen still shows up when you start searching for a word. If a user starts searching the screen should disappear.

https://user-images.githubusercontent.com/16003285/233020992-d28c4a88-c05a-45a0-8803-b887bc6ee771.mp4

### Urban dictionary result

I found the urban dictionary detail view to be a bit weird as the name gets repeated many times. Why is that and is it possible  to make it less redundant?

CleanShot 2023-04-19 at 10 52 17@2x

**Comment by DrChai** [comment]
@thomaslombart 

> The empty screen still shows up when you start searching for a word. If a user starts searching the screen should disappear.

It appears that the default empty view will always render as long as the list is empty, so I tried to prevent my custom empty view from rendering when the query is empty. However, the default view is still being rendered.
```typescript
<List>
...

  {
        query?null:<List.EmptyView title={emptyTitle} description={emptyDescription} />
  }
      
 </List>
```
https://user-images.githubusercontent.com/1713574/233427128-12d6e170-5a30-4b5b-b8bd-6e1e70dd2b58.mov

Although it is not displayed when `isLoading`, here I applied a debounce on input so fetch only starts when the user stop typing. Is there another way to control the empty view?

> I found the urban dictionary detail view to be a bit weird as the name gets repeated many times. Why is that and is it possible to make it less redundant?

Urban Dictionary always shows multiple definitions that users submitted for each term. So basically it has two endpoints, one for searching and another for browsing the specific term. Here is how it's actually implemented on its website:
Screenshot 2023-04-20 at 12 42 06 PM
Screenshot 2023-04-20 at 12 41 32 PM

>The title says The word(s) will be translated to 'en' when using Urban Dictionary. As far as it looks, there is no translation with the urban dictionary, right?

Yes, this message is a generic template and is shown on whatever the dictionary selected, I will submit a new commit to allow some dictionaries like Urban to have their own messages.

**Comment by thomaslombart** [comment]
> Although it is not displayed when isLoading, here I applied a debounce on input so fetch only starts when the user stop typing. Is there another way to control the empty view?

Using `isLoading` is the way to go to for that. The no results screen won't show while `isLoading` is `true`.

**Comment by DrChai** [comment]
> > Although it is not displayed when isLoading, here I applied a debounce on input so fetch only starts when the user stop typing. Is there another way to control the empty view?
> 
> Using `isLoading` is the way to go to for that. The no results screen won't show while `isLoading` is `true`.

Hi @thomaslombart, I think I got an unexpected result. As you can see from the screenshot, even when `isLoading` is updated to `true` and the loading bar pops up, the empty view is still present.
Screenshot 2023-04-21 at 8 10 57 AM
For me, the empty view only dismisses when the isLoading variable is set to true initially.

**Comment by thomaslombart** [comment]
That's happening because the component got rendered initially with `isLoading` set to `false.. To prevent that, you need to make sure you on the first render that `isLoading` is set to `true`.

**Review by thomaslombart — APPROVED** [review]
Looks good to me! As said in direct messages, there are certain patterns that could be improved but let's see if users actually have feedback about them 😊

**Pull request merged**